File: python/121.py
#121. You are given an array prices where prices[i] is the price of a given stock on the ith day.
# You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
# Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

from typing import List

def maxProfit(prices: List[int]) -> int:
    max_price, min_price = 0, float('inf')
    for price in prices:
        min_price = min(price, min_price)
        max_price = max(price - min_price, max_price)
    return max_price

prices = [int(x) for x in input().split()]
print(maxProfit(prices))

File: golang/121.go
package main

import (
	"fmt"
	"math"
)

func main() {
	fmt.Println(maxProfit([]int{7, 1, 5, 3, 6, 4}))
	fmt.Println(maxProfit([]int{7, 6, 4, 3, 1}))
}

// DESC: find the max profit by choosing a single day to buy a stock and
// choosing a different day in the future to sell that stock
func maxProfit(prices []int) int {
	max := float64(0)
	min := float64(1 << 62)
	for _, price := range prices {
		min = math.Min(float64(price), min)
		max = math.Max(float64(price)-min, max)
	}
	return int(max)
}
